### ISSUE TITLE
Why use "sigmoid" activation in MDense Layer?

### ISSUE THREAD

**Issue by BridgetteSong** [mozilla/LPCNet#164]
Why use _**sigmoid**_ activation not _**softmax**_ activation? if I don't use _**end2end**_ flag, loss function is **_sparse_categorical_crossentropy_**, and don't use **_from_logits_=True**, so _**y_pred**_ will not encode a probability distribution.

**Comment by BridgetteSong** [comment]
I found author committed codes as following, but i don't understand:
"_Saves on the MDense/softmax computation since we only need to compute 8 values instead of 256._"
[https://github.com/mozilla/LPCNet/commit/d24f49e346b85b5cb2c89c6dadcc913ce004fe83](url)

**Comment by BridgetteSong** [comment]
understand now

**Comment by vvfd87** [comment]
@BridgetteSong 
I'm really sorry to bothering you, but could you explain about "Representing output pdf as binary probability tree" commit?

**Comment by BridgetteSong** [comment]
> @BridgetteSong I'm really sorry to bothering you, but could you explain about "Representing output pdf as binary probability tree" commit?
> 
> I apply this commit, but synthesis speech quality is not good. (Synthesis speed is slightly improved.)
> 
> Using softmax and sample_from_pdf() instead of sample_mdense(), synthesis speech quality is good.
> 
> Do you have any idea?

1. author uses Hierarchical Sampling likes bit bunching, so 256 outputs can be splitted to 8 bit.
2. every bit has only two outputs(0 or 1), so we can use a sigmoid as activation function.
3. according to every bit output, we calculate final result.
